### ISSUE TITLE
Getting 2 rows of data when pasting to one row

### ISSUE THREAD

**Issue by joaopiopedreira** [awsp/handsontable-meteor#17]
Hi,

I installed your package and produced a very simple table (code below).

Every time I paste data into the spare row, I get 2 rows of data (with the data I've pasted). 

When I log the paste event `change` array (in the `afterChange` function below), I obtain an array with 2 arrays (same data in both, only the `id` is different.

Could you point me in the right direction?

My complete code below:

```
Template.dataManager.onRendered(function(){
    Session.set(SHOW_SPINNER,false);

    var meteorData = [];

    var container = document.getElementById('data-manager-content');

    var hot = new Handsontable(container, {
        readOnly:false,
        data: meteorData,
        height: 396,
        minSpareRows: 1,
        fixedRowsTop: 1,
        manualColumnResize: true,
        colHeaders: [
            'Customer',
            'Document No.',
            'Date',
            'Phone',
            'Address',
            'Postal Code',
            'Vehicle',
            'Closed',
            'Updated'],
        rowHeaders:true,
        autoWrapRow: true,
        //colWidths: [200, 85, 70, 70, 70, 70, 70, 70],
        contextMenu: true,
        beforeRemoveRow: function(){return confirm("Are you sure you want to remove this row?")},
        columns: [
            {data: 'customer'},
            {data: 'docNo'},
            {data: 'date',type:'date',dateFormat: 'YYYY-MM-DD'},
            {data: 'contactPhone'},
            {data: 'address'},
            {data: 'postCode'},
            {data: "name",
                type: 'dropdown',
                source: _.uniq(DumbLists.find({},{fields:{name:1}}).map(function(it){return it.name}))
            },
            {data: 'closed',type:'checkbox',readOnly:true},
            {data: 'confirmClose',type:'checkbox',readOnly:true}
        ],
        afterChange: function (change, source) {  // "change" is an array of arrays.
            if (source === "edit") {  // Don't need to run this when data is loaded
                console.log(change);

                for (var i = 0; i < change.length; i++) {   // For each change, get the change info and update the record
                    var rowNum = change[i][0]; // Which row it appears on Handsontable
                    var row = meteorData[rowNum];  // Now we have the whole row of data, including _id
                    var key = change[i][1];  // Handsontable docs calls this "prop"
                    var oldVal = change[i][2];
                    var newVal = change[i][3];
                    var setModifier = {$set: {}};   // Need to build $set object
                    var randomId = Random.id();
                    var companyId = row.companyId;
                    var vehicleId = row.driverId;
                    setModifier.$set[key] = newVal; // So that we can assign 'key' dynamically using bracket notation of JavaScript object
                    if(row._id){
                        Meteor.call('db_update','dumb_lists',row._id,setModifier,userId || Meteor.userId(),companyId,vehicleId,randomId);
                    } else {
                        //TODO: Logic to insert new record in Meteor from row
                    }
                }
            } else
            if (source !== "paste") {
                console.log(change);
            } else {
                console.log(change, source);
            }
        }
    });

    Tracker.autorun( function () {
        meteorData = DumbLists.find().fetch();
        hot.loadData(meteorData);
    });


});
```

**Comment by awsp** [comment]
Not sure if I fully understand what you were trying to say. 
So you are logging the "paste" event, and you are getting 2 rows back when you paste only 1 row?

As far as I could tell in the code, you haven't handled the paste event, so I assume no database insertions yet done. (Correct me if I am wrong). I tried to replicate similar situation like yours, so far I do not see any errors or misbehavior. Would that be possible you could replicate the result on http://meteorpad.com/ so I could take a look in more detail?

**Comment by joaopiopedreira** [comment]
Hi @awsp, it's really strange.

I've copy/pasted my code into [meteorpad](http://meteorpad.com/pad/NrRyYJBjxPWzKArSA/HandsOnTable_issue#17) and the problem is not showing.

Could you tell me what should I look for (events, data, etc.) in order to debug this issue in my code? Thanks.

**Comment by awsp** [comment]
@joaopiopedreira Sorry just back from the long holidays. 
Your meteorpad does look okay to me too. Console shows correct result. 
So I don't really know why you are experiencing the issue on your local machine.

Okay, couple things we can try here. 
I notice you are using the package version 0.10.0, I would suggest try update to the latest version of awsp:handsontable so you could get the latest of Handsontable and see it solves the problem for you. If it still doesn't, try inspect it from Chrome's debugger and observe the afterChange event. 

One other thing tho, which I am currently suspecting, is that on the Tracker loop when using the autopublish package, if you use it incorrectly, random things could happen due to some sort of race condition. (Forgive me if I use a wrong term here) So basically, the collection doesn't clean up before DDP resends all the items back from the database. Try switch to the pub/sub way instead and see if it solves the problem. So instead of reloading the whole collection, observe the added, changed, removed events and modify handsontable.

**Comment by joaopiopedreira** [comment]
Thank you for your reply.
I'll update the package and see if it solves the issue.
As to the second part, it's not a tracker issue. I've tried it with a a simple
js array as data (no meteor collection) and the problem persists.
Well, we'll get there Again, thanks.

On 1 December 2015 at 06:44, Anthony S. Wu < notifications@github.com > wrote:
@joaopiopedreira Sorry just back from the long holidays.
Your meteorpad does look okay to me too. Console shows correct result.
So I don't really know why you are experiencing the issue on your local machine.

Okay, couple things we can try here.
I notice you are using the package version 0.10.0, I would suggest try update to
the latest version of awsp:handsontable so you could get the latest of
Handsontable and see it solves the problem for you. If it still doesn't, try
inspect it from Chrome's debugger and observe the afterChange event.

One other thing tho, which I am currently suspecting, is that on the Tracker
loop when using the autopublish package, if you use it incorrectly, random
things could happen due to some sort of race condition. (Forgive me if I use a
wrong term here) So basically, the collection doesn't clean up before DDP
resends all the items back from the database. Try switch to the pub/sub way
instead and see if it solves the problem. So instead of reloading the whole
collection, observe the added, changed, removed events and modify handsontable.

**Comment by joaopiopedreira** [comment]
Hi @awsp,

For future reference:

I created a virgin meteor app in this [simple github repo](https://github.com/joaopiopedreira/meteor-handsOnTable-issue_17). Then I [patiently] added my packages one by one (I have maaany packages...). 

It turns out that the clashing package is [meteorhacks:zones](https://github.com/meteorhacks/zones). If you add this package, you have the double paste issue. I removed the package from my project and it all works like a charm now.

It would be interesting to find out why this clash occurs...
